### DISCUSSION TITLE
how to do a joined where in prisma query?

### DISCUSSION THREAD

**Discussion by adrian-goe** [prisma/prisma#8983] [Q&A]
Hi, I have this simple SQL query 
```sql
SELECT s.id, s.name, s.logo_url
  FROM supplier AS s
  JOIN supplier_user su ON s.id = su."supplierId"
 WHERE su."userId" = ${id}
   AND su.active = TRUE
```
I tried to build this with a prisma query, but I don’t get this to work. May last try was:
```ts
return this.prisma.supplier.findMany({
  where: {
    supplierUser: {
      some: {
        active: true,
        userId: { equals: id }
      }
    }
  }, select: {
    id: true,
    name: true,
    logoUrl: true,
  }
});
```
Can someone please help me to understand how this joined selects work?

**Comment by adrian-goe** [comment]
I changed the query to this:
```ts
where: {
  supplierUser: {
    some: {
      AND: [
        { active: true },
        { userId: { equals: id } }
      ]
    }
  }
},
select: {
  id: true,
  name: true,
  logoUrl: true
}
```

What I noticed now is, that the generated SQL from the debug interprets the `{ userId: { equals: id } }` as `"t0"."id" IS NOT NULL` and not as `"t0"."id" = 'some input'`. Idk how to fix this.

```sql
SELECT "public"."supplier"."id", "public"."supplier"."name", "public"."supplier"."logo_url"
FROM "public"."supplier"
WHERE ("public"."supplier"."id") IN (SELECT "t0"."id"
                                     FROM "public"."supplier" AS "t0"
                                              INNER JOIN "public"."supplier_user" AS "j0" ON ("j0"."supplierId") = ("t0"."id")
                                     WHERE ("j0"."active" = true AND "t0"."id" IS NOT NULL));
```

**Reply by ryands17** [reply]
This should work:

```ts
where: {
  supplierUser: {
     AND: [
      { some: { active: true } },
      { some:  { userId: id } }
     ]
  }
},
select: {
  id: true,
  name: true,
  logoUrl: true
}
```

**Comment by adrian-goe** [comment]
Unfortunately, this does not work directly.
![Screenshot 2021-08-31 at 12 34 36](https://user-images.githubusercontent.com/17064122/131487872-31a16a75-16e0-46ef-b427-603965c2ffa0.png)

**Reply by ryands17** [reply]
Ahh, the `AND` needs to go on top then:

```ts
where: {
     AND: [
      { supplierUser: { some: { active: true } } },
      { supplierUser: { some:  { userId: id } } }
     ]
},
select: {
  id: true,
  name: true,
  logoUrl: true
}
```

**Reply by adrian-goe** [reply]
This results in this following sql, the userId is still missing. 🙁

```sql
SELECT "public"."supplier"."id", "public"."supplier"."name", "public"."supplier"."logo_url"
FROM "public"."supplier"
WHERE (("public"."supplier"."id") IN (SELECT "t0"."id"
                                      FROM "public"."supplier" AS "t0"
                                               INNER JOIN "public"."supplier_user" AS "j0" ON ("j0"."supplierId") = ("t0"."id")
                                      WHERE ("j0"."active" = $1 AND "t0"."id" IS NOT NULL)) AND
       ("public"."supplier"."id") IN (SELECT "t0"."id"
                                      FROM "public"."supplier" AS "t0"
                                               INNER JOIN "public"."supplier_user" AS "j0" ON ("j0"."supplierId") = ("t0"."id")
                                      WHERE (1 = 1 AND "t0"."id" IS NOT NULL)))
OFFSET $2;
```

But i have this workaround for me:
```ts
where: {
  supplierUser: {
    some: {
      AND: [
        { active: true },
        { userId: `${id}` }
      ]
    }
  }
},
select: {
  id: true,
  name: true,
  logoUrl: true
}
```
The thing, I do not understand now is the explicit `string` cast for the `id` since the function param is already a string.

**Reply by adrian-goe** [reply]
there is one mor thing i dont get:
Why is there this line in the SQL?
```sql
WHERE (1 = 1 AND "t0"."id" IS NOT NULL))
```
`1=1` is always true. so why check this?

**Reply by ryands17** [reply]
I would like to have your schema so that I can verify with sample data. You should get the correct results with this.

As for the `where 1 = 1`, the reason for this is it's easier to build multiple conditions programmatically. Have a look at this [ SO answer](https://stackoverflow.com/questions/1264681/what-is-the-purpose-of-using-where-1-1-in-sql-statements)

**Reply by adrian-goe** [reply]
The full schema, prisma versions and so are here: https://github.com/prisma/prisma/issues/8992

Thx, I will look into this.

**Reply by ryands17** [reply]
Works totally fine and I get the `userId` perfectly.

![image](https://user-images.githubusercontent.com/19697099/131492871-703c6583-cecd-4676-a6e6-e6d5415ffa7f.png)


Here's the code:
[test-prisma.zip](https://github.com/prisma/prisma/files/7083428/test-prisma.zip)

**Reply by adrian-goe** [reply]
Yes thx, this query works.
`some: { userId: `${id}` },` or `some: { userId: '2290beaf-530a-4fca-8d6c-7bd3d856cdaf' },` 

But `some: { userId: id },` not. 
And that, although ID already has the type `string`. `public findSupplierByUserId(id: string): Promise<any> {`
Somehow I need an explicit string type cast when this is already type if string 🙈  and this will not go into my head. Maybe it's a TS problem and not related to prisma at all.

**Reply by ryands17** [reply]
Seems so then. Logging `typeof id` should give you what's the type of parameter being passed. I tried checking by adding a variable to a function and it still works. So possibly a TS or variable passing configuration issue then.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
